scholly/service/core/push.js

```diff
@@ -0,0 +1,186 @@
+ */
+function cleanupBuild() {
+  fs.rm(`${PATH_TO_BASE_REACT_APP}/build`, { recursive: true }, (err) => {
+    if (err) {
+      winston.error('Failed to remove built directory.');
+    }
+  });
+}
+
+async function makeGHRepo(ghUsername, ghToken) {
+  // check if the repo exists
+  const repoName = ghUsername + '.github.io';
+  const searchQuery = 'q=user:' + ghUsername + ' ' + repoName + ' in:name';
+  const options = {
+    Authorization: 'token ' + ghToken,
+    Accept: 'application/vnd.github.v3+json',
+  };
+  let repoResponse;
+  try {
+    repoResponse = await axios.get(
+      'https://api.github.com/search/repositories?' + searchQuery,
+      {
+        headers: options,
+      }
+    );
+  } catch (err) {
+    console.log(err);
+  }
+
+  if (repoResponse.data.total_count === 1) {
+    console.log('repo exists already');
+  } else {
+    // if it doesn't, make one
+    const createRepoBody = {
+      name: repoName,
+      private: false, // free accounts can only use a public repo for GH pages
+      auto_init: true,
+    };
+    try {
+      await axios({
+        url: 'https://api.github.com/user/repos',
+        method: 'post',
+        headers: options,
+        data: createRepoBody,
+      });
+      console.log('Repo was created successfully');
+    } catch (err) {
+      console.log(err);
+    }
+  }
+
+  // set up pages for the repo regardless of whether its new or not
+  // see https://docs.github.com/en/rest/reference/repos#pages
+  try {
+    const pagesResponse = await axios.get(
+      'https://api.github.com/repos/' + ghUsername + '/' + repoName + '/pages',
+      {
+        headers: {
+          Authorization: 'token ' + ghToken,
+          Accept: 'application/vnd.github.switcheroo-preview+json',
+        },
+      }
+    );
+    console.log('GH Pages already configured for repo');
+  } catch (err) {
+    console.log(err);
+    console.log('GH pages not created yet');
+    createPagesSite(ghUsername, ghToken);
+  }
+}
+
+async function createPagesSite(ghUsername, ghToken) {
+  const pagesBody = {
+    source: {
+      branch: 'main',
+      path: '/',
+    },
+  };
+  const repoName = ghUsername + '.github.io';
+  let createPagesResponse;
+  try {
+    createPagesResponse = await axios({
+      url:
+        'https://api.github.com/repos/' +
+        ghUsername +
+        '/' +
+        repoName +
+        '/pages',
+      method: 'post',
+      headers: {
+        Authorization: 'token ' + ghToken,
+        Accept: 'application/vnd.github.switcheroo-preview+json',
+      },
+      data: pagesBody,
+    });
+    if (createPagesResponse.status === 201) {
+      console.log('Pages site successfully created');
+    } else {
+      // see https://docs.github.com/en/rest/reference/repos#create-a-github-pages-site
+      console.log(createPagesResponse.data);
+    }
+  } catch (err) {
+    console.log(err);
+  }
+}
+
+async function buildPages(ghUsername, ghToken) {
```

Comment:
What does this func do? It seems like it builds the client website, but I thought that's automatically done by GitHub based off the repository's main branch.

Reply:
You're right about that - when you push to the main branch, it should automatically build the site, but I did notice straight after the push, if you go into settings of your repo -> pages, it says your site is ready to be published instead of your site is published, so this function is just to trigger a build on top of that to make sure it publishes, which takes a few seconds. I could just be impatient which is why I have this function, haha.